### ISSUE TITLE
[feature] --target argument for conan build

### ISSUE THREAD

**Issue by petermbauer** [conan-io/conan#6405]
Following the broader discussion in #5553 i suggest the following feature mainly relevant for CI but also for local development:
For doing "non-standard" builds like generating documentation output or running a static code analysis people usually create optional build targets in the used build system, e.g. CMake.

As there is most probably no need to include the generated files in the Conan packages, the build is done with `conan install && conan build` in the userspace to easily access the files for uploading, archiving, .... To avoid having to call the underlying build system directly in the CI builds (may also be used locally by developers so they do not have to run `activate.bat` and call CMake directly with all required parameters), the actual build target could be passed directly with e.g. `conan build --target docs` to the build() method, similar to the `--configure`, `--build`, `--test` flags and the respective attributes.

This would avoid having to add dedicated options for all these "non-standard" builds in the conanfile.py which are not used for package creation anyway.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by Minimonium** [comment]
In my opinion, the current `conan build` capabilities are very inflexible and fragile (working directories, imported files, etc) and require an overhaul. 
It's a good idea to think of a more generalized interface for package orchestration, something along the lines of the https://github.com/conan-io/conan/issues/5826. But it's quite a big task and I'm afraid not a priority right now.

**Comment by ghost** [comment]
I agree that Conan should provide enough common build system commands that it is possible to abstract most build system tasks.

I have complained about this 2 years ago in #2470 and #2499.

**Comment by DukeOfKirkcaldy** [comment]
I am also missing this feature, but also for `conan create` since our CI system is using the creator commands.

**Comment by petermbauer** [comment]
> I am also missing this feature, but also for `conan create` since our CI system is using the creator commands.

If you are really creating binary packages then these have to contain the produced files and should be distinguishable via options anyway. Please describe your use-case in more detail.

**Comment by DukeOfKirkcaldy** [comment]
> If you are really creating binary packages then these have to contain the produced files and should be distinguishable via options anyway. Please describe your use-case in more detail.

I thought my use case was the same as yours - to be able to build optional build targets without having to call the underlying build system itself.  For example, generating documentation for release builds.

But nevermind, I read more clearly and now I see your point.  I can do this through the recipe.

Thanks

**Comment by ghost** [comment]
After thinking about it for longer I'm not so sure anymore.

One of my use cases was that I've had a utility that used my library in the same repo. But upon further thinking, I think declaring several targets inside the build system and having dependencies between there is a bad idea and duplication of effort.

Each library should at least be in a separate directory with separate `conanfile.py` and, in case of CMake, `CMakeLists.txt`. Dependencies should only be declared in `conanfile.py`. Like so:

```
Library
    - include
    - src
    - conanfile.py
    - CMakeLists.txt
Utility
    - src
    - conanfile.txt/py
    - CMakeLists.txt
```

Of course, utility can be a Conan package too that will be run in build context when building other code.

This leaves me wondering if there is ever a valid use case for several targets in the build system.

**Comment by petermbauer** [comment]
> This leaves me wondering if there is ever a valid use case for several targets in the build system.

As stated above my use-case would be generated documentation and static code analysis which i think are still valid.

**Comment by petermbauer** [comment]
@jgsogo, any thoughts on this? Or are there plans/ideas for a more "generalized" approach?

**Comment by jgsogo** [comment]
Hi! No news about this.

Conan v2.0 will include a new command-line interface that makes it possible to add user-defined commands (some preliminary PRs are already merged). This new CLI together with a good documentation of the Conan API opens a new set of possibilities, but IMHO it's not enough for this `--target` argument/need.

I'm not sure if this is something desired for the `conan build` command or it is more related to the _editables_ feature (install the package as editable and then run local tools/commands). We are facing many issues with the local developer workflow not behaving exactly the same way as the `conan create` and it might not be the best idea to add more complexity (or flexibility) to it if it is not 100% needed.

Maybe this issue is a symptom of something different: we usually don't generate and package things like docs, benchmark results, static code analysis reports,... because we don't need this information always, we don't want to add more options and generate different packages with and without this extra files. If this is the symptom, it is more related to the ongoing discussions about quality artifacts, PDBs,... and we are thinking a way to handle this files and add them to the package if available without affecting the package-id.

**Comment by markovandooren** [comment]
@jgsogo Would the new API allow me to let me call a particular piece of code in my conanfile.py after setting up the dependency information?

I just managed to get Visual Studio source indexing to work for Conan packages with source coming from various repositories. But the developers are working in Visual Studio. For DLLs, there is no problem, but Visual Studio cannot merge source indexing from PDBs of static libraries. That information is required to be able to step into code from the static library when debugging. The conanfile.py contains the dependency information and logic to merge the source indexing for such PDBs. All I need for a seemless Visual Studio workflow is to trigger the Conan source indexing code from Visual Studio. This can be done in MSBuild with a custom Target where I execute 'conan some_command .' 

There is however no appropriate command to execute. The 'build' command is not appropriate because the sources have just been built by Visual Studio. The 'package' command is not appropriate because I don't want to copy large files after every compilation, and I'm working in editable mode anyway. 

If I could just call the index_sources() method on the ConanFile object after all the dependency information has been processed, then that would be ideal. My only option now seems to be to abuse the 'build' or 'package' command and make it behave a bit different e.g. when it is it not in the local cache. To be able to still debug my package I would have to build in a mechanism to still let it function normally (e.g. by checking if conan_debug.txt exists).

A local developer workflow will do a few things that Conan does already, such as building the sources. That means that in order to get the same end result, we have to be able to configure our tools to let Conan do the other steps.

**Comment by jgsogo** [comment]
Hi, @markovandooren. Your request is very specific and it probably deserves its own issue (to be assigned to CLI and 2.0 version). We can have a look into it and keep it in mind while working in the Conan API. Please, be as explicit as possible, we need to know the context while running that _custom_ recipe method: values accessible within this method (settings, environment variables, `user_info`, folders,...), should the API provide methods to activate/deactivate these contexts? should all _custom commands_ run with the same context? ...? are there any limitations/functionality required to it?

**Comment by markovandooren** [comment]
Hi @jgsogo, I will create a new detailed issue next week. I had to make a change to the mechanism to avoid a problem with concurrent builds in Visual Studio. It affected the way that I interact with Conan, so I want to do more testing to make sure that my workflow covers everything.

**Comment by petermbauer** [comment]
> Maybe this issue is a symptom of something different: we usually don't generate and package things like docs, benchmark results, static code analysis reports,... because we don't need this information always, we don't want to add more options and generate different packages with and without this extra files. If this is the symptom, it is more related to the ongoing discussions about quality artifacts, PDBs,... and we are thinking a way to handle this files and add them to the package if available without affecting the package-id.

Not sure if i got this right. As it is now one can either add the files to a package without introducing an additional option (-> same package_id) or with an option to have it in a separate binary package or even both. The point is that the option is a way to decide what should be built and added to the package but if the files shall not be added to the package but just archived or processed in some other way there is no way to "communicate" that from the commandline to the `build()` method so the user or the CI has to move down to the level of the build system to perform the respective build.

Like the `conan build` options `--configure`, `--build` and `--test` such a `--target` argument would not be used for regular package creation but only for local development so i understand that these are not core features of a package manager but highly appreciated in some cases and IMHO the maintenance effort is not that high.

**Comment by petermbauer** [comment]
The following workaround came to my mind, so far this approach looks promising though it requires an additional e.g. `conan install -o build_target=coverage ..` in case a different target shall be built but this is not a big deal for the CI. And even if somebody would create a binary package from such a special build target, it would be clearly visible in the package.
```
class MyPackage(ConanFile):
    ...
    # option for building special targets directly with Conan
    options = {"build_target": "ANY"}
    default_options = {"build_target": None}
    ...
    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build(target=None if not self.options.build_target else str(self.options.build_target))
    ...
```
Are there any other risks/drawbacks i missed?

**Comment by RolfBippus** [comment]
First of all, I would also very much like to have such a feature.
@petermbauer 👍 
I have used what you proposed, however with one added thing I found important. 

In order to prevent the additional options from changing the package id (specifying a specific target should not change that, since the settings remain the same) I used the default value (according to the conan docs of [package_id()](https://docs.conan.io/en/latest/reference/conanfile/methods.html#package-id) )
```
def package_id(self):
        # "deactivate" options that are irrelevant for the package ID 
        # see https://docs.conan.io/en/latest/reference/conanfile/methods.html#package-id
        self.info.options.build_target = self.default_options["build_target"]
```
Trying to simply remove the build_target attribute from self.info.options did not work, since then the option seems to be gone completely, even from self.options (I did not test this for a while... might also be some flaw in my code...)

**Comment by petermbauer** [comment]
Thanks for sharing @RolfBippus, i also considered removing it from the package_id but for me its ok to produce different binary packages as the content depends on which target is built.

**Comment by memsharded** [comment]
Closing the ticket as old and not planned at the moment.
